File: plugins/plugin-instagram/typescript/src/providers/user-state.ts
/**
 * User state provider for Instagram
 */

import type {
  IAgentRuntime,
  Memory,
  Provider,
  ProviderResult,
  State,
} from "@elizaos/core";

/**
 * Provider for Instagram user state
 *
 * Provides contextual information about the current Instagram interaction,
 * including user ID, thread ID, and whether the context is a DM or comment.
 */
export const userStateProvider: Provider = {
  name: "instagram_user_state",
  description:
    "Provides Instagram user context state including user ID, thread ID, and interaction type",

  async get(
    _runtime: IAgentRuntime,
    message: Memory,
    _state: State,
  ): Promise<ProviderResult> {
    const content = message.content as Record<string, unknown>;

    // Extract Instagram context from the message
    const userId = content.userId as number | undefined;
    const threadId = content.threadId as string | undefined;
    const mediaId = content.mediaId as number | undefined;
    const roomId = message.roomId;

    const isDm = threadId !== undefined && threadId !== null;
    const isComment = mediaId !== undefined && mediaId !== null;

    const stateData = {
      user_id: userId ?? null,
      thread_id: threadId ?? null,
      media_id: mediaId ?? null,
      room_id: roomId ?? null,
      is_dm: isDm,
      is_comment: isComment,
    };

    return {
      text: JSON.stringify(stateData, null, 2),
      values: {
        instagram_user_id: userId ?? "",
        instagram_thread_id: threadId ?? "",
        instagram_media_id: mediaId ?? "",
        instagram_is_dm: isDm,
        instagram_is_comment: isComment,
      },
      data: stateData,
    };
  },
};

File: plugins/plugin-instagram/rust/src/providers/user_state.rs
//! User state provider for Instagram

use async_trait::async_trait;
use serde_json::Value;

use super::{InstagramProvider, ProviderContext};

/// Provider for Instagram user state
pub struct UserStateProvider;

#[async_trait]
impl InstagramProvider for UserStateProvider {
    fn name(&self) -> &'static str {
        "instagram_user_state"
    }

    async fn get(&self, context: &ProviderContext) -> Value {
        serde_json::json!({
            "user_id": context.user_id,
            "thread_id": context.thread_id,
            "media_id": context.media_id,
            "room_id": context.room_id,
            "is_dm": context.thread_id.is_some(),
            "is_comment": context.media_id.is_some()
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_user_state_dm() {
        let provider = UserStateProvider;
        let context = ProviderContext {
            user_id: Some(12345),
            thread_id: Some("thread-1".to_string()),
            media_id: None,
            room_id: Some("room-uuid".to_string()),
        };

        let state = provider.get(&context).await;
        assert_eq!(state["user_id"], 12345);
        assert_eq!(state["is_dm"], true);
        assert_eq!(state["is_comment"], false);
    }

    #[tokio::test]
    async fn test_user_state_comment() {
        let provider = UserStateProvider;
        let context = ProviderContext {
            user_id: Some(12345),
            thread_id: None,
            media_id: Some(67890),
            room_id: Some("room-uuid".to_string()),
        };

        let state = provider.get(&context).await;
        assert_eq!(state["user_id"], 12345);
        assert_eq!(state["is_dm"], false);
        assert_eq!(state["is_comment"], true);
    }
}
